moveit_core/robot_model/src/joint_model_group.cpp

```diff
@@ -97,188 +101,231 @@ bool jointPrecedes(const JointModel* a, const JointModel* b)
-      group_mimic_update_.push_back(mu);
-    }
+  const auto is_ascending = [](const auto& sub_view) {
+    const auto& values = sub_view | ranges::to<std::vector>();
+    return values[0] + 1 == values[1];
+  };
+  is_contiguous_index_list_ = !variable_index_list_.empty()
+    && ranges::all_of(variable_index_list_ | views::sliding(2), is_ascending);
+
+  // when updating/sampling a group state only,
+  // only mimic joints that have their parent within the group get updated.
+  group_mimic_update_ = mimic_joints_
+    | views::filter([this](const auto& jm) { return hasJointModel(jm->getMimic()->getName()); })
+    | views::transform([this](const auto& jm) {
+      int src = joint_variables_index_map_[jm->getMimic()->getName()];
+      int dest = joint_variables_index_map_[jm->getName()];
+      return GroupMimicUpdate(src, dest, jm->getMimicFactor(), jm->getMimicOffset());
+    })
+    | ranges::to<std::vector>();
 
   // now we need to make another pass for group links (we include the fixed joints here)
-  std::set<const LinkModel*> group_links_set;
-  for (const JointModel* joint_model : joint_model_vector_)
-    group_links_set.insert(joint_model->getChildLinkModel());
-  for (const LinkModel* group_link : group_links_set)
-    link_model_vector_.push_back(group_link);
-  std::sort(link_model_vector_.begin(), link_model_vector_.end(), OrderLinksByIndex());
-
-  for (const LinkModel* link_model : link_model_vector_)
-  {
-    link_model_map_[link_model->getName()] = link_model;
-    link_model_name_vector_.push_back(link_model->getName());
-    if (!link_model->getShapes().empty())
-    {
-      link_model_with_geometry_vector_.push_back(link_model);
-      link_model_with_geometry_name_vector_.push_back(link_model->getName());
-    }
-  }
+  link_model_vector_ = joint_model_vector_
+    | views::transform([](const auto& jm) { return jm->getChildLinkModel(); })
+    | views::unique
+    | ranges::to<std::vector>()
+    | actions::sort(OrderLinksByIndex());
+  link_model_map_ = link_model_vector_
+    | views::transform([](const auto& lm) { return std::make_pair(lm->getName(), lm); })
+    | ranges::to<std::map>();
+  link_model_name_vector_ = link_model_vector_
+    | views::transform(getElementName)
+    | ranges::to<std::vector>();
+  link_model_with_geometry_vector_ = link_model_vector_
+    | views::filter(has_geometry)
+    | ranges::to<std::vector>();
+  link_model_with_geometry_name_vector_ = link_model_with_geometry_vector_
+    | views::transform(getElementName)
+    | ranges::to<std::vector>();
 
   // compute the common root of this group
   if (!joint_roots_.empty())
   {
-    common_root_ = joint_roots_[0];
-    for (std::size_t i = 1; i < joint_roots_.size(); ++i)
-      common_root_ = parent_model->getCommonRoot(joint_roots_[i], common_root_);
+    common_root_ = ranges::accumulate(joint_roots_, joint_roots_[0],
+      [this](const auto& left, const auto& right) {
```

Comment:
See, this. It feels like this should be possible for all the zipped views, instead of having to use `std::get` or unpack a parameter.

Reply:
I agree and I've tried but am unable to get it to compile.  accumulate specifically takes a binary function.